lang/elaborator/src/normalizer/read_back.rs

```diff
@@ -56,31 +56,32 @@ impl ReadBack for val::Neu {
 
     fn read_back(&self, prg: &ust::Prg) -> Result<Self::Nf, TypeError> {
         let res = match self {
-            val::Neu::Var { span, name, idx } => ust::Exp::Variable(ust::Variable {
+            val::Neu::Var { span, name, idx } => ust::Exp::Variable(Variable {
                 span: *span,
-                info: (),
-                ctx: None,
-                name: name.clone(),
                 idx: *idx,
+                name: name.clone(),
+                inferred_type: None,
             }),
             val::Neu::Dtor { span, exp, name, args } => ust::Exp::DotCall(ust::DotCall {
                 span: *span,
-                info: (),
                 exp: exp.read_back(prg)?,
                 name: name.clone(),
                 args: ust::Args { args: args.read_back(prg)? },
+                inferred_type: None,
             }),
             val::Neu::Match { span, name, on_exp, body } => ust::Exp::LocalMatch(ust::LocalMatch {
                 span: *span,
-                info: (),
                 ctx: None,
                 motive: None,
                 ret_typ: None,
                 name: name.clone(),
                 on_exp: on_exp.read_back(prg)?,
                 body: body.read_back(prg)?,
+                inferred_type: None,
             }),
-            val::Neu::Hole { span } => ust::Exp::Hole(ust::Hole { span: *span, info: () }),
+            val::Neu::Hole { span } => {
+                ust::Exp::Hole(Hole { span: *span, inferred_type: None, inferred_ctx: None })
```

Comment:
Just a suggestion: embedding into `Exp` can be made a bit more convenient by creating an implementation of `Into<Exp>` for all expression structs. Instead of `ust::Exp::Hole(Hole { ... })` you could then write `Hole { ... }.into()`.

Reply:
Good idea! I added a `From` instance for all expression structs so that we can use this in the future.